components/scream/src/physics/p3/p3_functions_rain_sed_impl.hpp

```diff
@@ -35,6 +33,121 @@ ::compute_rain_fall_velocity(
+  const uview_1d<Spack>& nr,
+  const uview_1d<Spack>& nr_incld,
+  const uview_1d<Spack>& mu_r,
+  const uview_1d<Spack>& lamr,
+  const uview_1d<Spack>& rflx,
+  const uview_1d<Spack>& qr_tend,
+  const uview_1d<Spack>& nr_tend,
+  Scalar& prt_liq)
+{
+  // Get temporary workspaces needed for the ice-sed calculation
+  uview_1d<Spack> V_qr, V_nr, flux_qx, flux_nx;
+  workspace.template take_many_contiguous_unsafe<4>(
+    {"V_qr", "V_nr", "flux_qx", "flux_nx"},
+    {&V_qr, &V_nr, &flux_qx, &flux_nx});
+
+  const view_1d_ptr_array<Spack, 2>
+    fluxes_ptr = {&flux_qx, &flux_nx},
+    vs_ptr     = {&V_qr, &V_nr},
+    qnr_ptr    = {&qr, &nr};
+
+  const auto sflux_qx = scalarize(flux_qx);
+
+  // find top, determine qxpresent
+  const auto sqr = scalarize(qr);
+  constexpr Scalar qsmall = C::QSMALL;
+  bool log_qxpresent;
+  const Int k_qxtop = find_top(team, sqr, qsmall, kbot, ktop, kdir, log_qxpresent);
+
+  if (log_qxpresent) {
+    Scalar dt_left   = dt;  // time remaining for sedi over full model (mp) time step
+    Scalar prt_accum = 0.0; // precip rate for individual category
+
+    // find bottom
+    Int k_qxbot = find_bottom(team, sqr, qsmall, kbot, k_qxtop, kdir, log_qxpresent);
+
+    while (dt_left > C::dt_left_tol) {
+      Scalar Co_max = 0.0;
+      Int kmin, kmax;
+      const Int kmin_scalar = ( kdir == 1 ? k_qxbot : k_qxtop);
+      const Int kmax_scalar = ( kdir == 1 ? k_qxtop : k_qxbot);
+
+      Kokkos::parallel_for(
+       Kokkos::TeamThreadRange(team, V_qr.extent(0)), [&] (Int k) {
+        V_qr(k) = 0;
+        V_nr(k) = 0;
+      });
+      team.team_barrier();
+
+      // Convert top/bot to pack indices
+      util::set_min_max(k_qxbot, k_qxtop, kmin, kmax, Spack::n);
+
+      // compute Vq, Vn (get values from lookup table)
+      Kokkos::parallel_reduce(
+       Kokkos::TeamThreadRange(team, kmax-kmin+1), [&] (int pk_, Scalar& lmax) {
+
+        const int pk = kmin + pk_;
+        const auto range_pack = scream::pack::range<IntSmallPack>(pk*Spack::n);
+        const auto range_mask = range_pack >= kmin_scalar && range_pack <= kmax_scalar;
+        const auto qr_gt_small = range_mask && qr_incld(pk) > qsmall;
+        if (qr_gt_small.any()) {
+          compute_rain_fall_velocity(qr_gt_small, vn_table, vm_table,
+                                     qr_incld(pk), rcldm(pk), rhofacr(pk),
+                                     nr(pk), nr_incld(pk), mu_r(pk), lamr(pk), V_qr(pk), V_nr(pk));
+        }
+        const auto Co_max_local = max(qr_gt_small, -1,
+                                      V_qr(pk) * dt_left * inv_dzq(pk));
+        if (Co_max_local > lmax) lmax = Co_max_local;
+      }, Kokkos::Max<Scalar>(Co_max));
+      team.team_barrier();
+
+      generalized_sedimentation<2>(rho, inv_rho, inv_dzq, team, nk, k_qxtop, k_qxbot, kbot, kdir, Co_max, dt_left, prt_accum, fluxes_ptr, vs_ptr, qnr_ptr);
+
+      // AaronDonahue, rflx output
+      util::set_min_max(k_qxbot+1, k_qxtop+1, kmin, kmax, Spack::n);
```

Comment:
@ambrad , there was a change in the reference rain_sedimentation routine since we did the micro app. The main change was the addition of

```
         !-- AaronDonahue, rflx output                                                                                                                                                                                                         
         do k = k_qxbot,k_qxtop,kdir
            rflx(k+1) = rflx(k+1) + flux_qx(k) ! AaronDonahue                                                                                                                                                                                  
         enddo
```
... which I implemented here (lines 124-132). I tried to minimize scalarization, but this implementation does require a scalarized flux_qx.